### PR TITLE
Implement Camera Multiplexing API

### LINKED ISSUE #12077: [WIP] Proposal: camera multiplexing API
# OpenCV camera multiplexing API

## Motivation
Current `cv::VideoCapture` provide only blocking-style API for access cameras, 
call to `VideoCapture::read` or `VideoCapture::grab` will block calling thread until
frame data is available.
When working with multiple cameras this require to create dedicated thread for each
camera.

We propose a way to wait for multiple cameras in single thread and a method to check
for camera frame availability without blocking calling thread.

## New OpenCV APIs

"videoio.hpp":
```C++

enum GrabAnyResult {
  GRAB_ANY_TIMEOUT,
  GRAB_ANY_EVENT,
  GRAB_ANY_READY
}

class GrabAnyEvent {
public:
  GrabAnyEvent();
  GrabAnyEvent(const GrabAnyEvent&);
  GrabAnyEvent& operator=(const GrabAnyEvent&);

  void signal();
  void reset();
};

struct GrabAnyItem {
  VideoCapture* capture;
  bool ready;
};

GrabAnyResult grabAny(/*some_array_type<GrabAnyItem>*/ items, GrabAnyEvent* event, int timeout_milliseconds);
```

#### Parameters
* items - array of GrabAnyItem structures (can be implicitly created from `std::vector<GrabAnyItem>`, `std::array<GrabAnyItem>`, or `GrabAnyItem[]`),
used both as input and output.
User must set `capture` pointer to valid, opened `VideoCapture` instance.
After return `ready` field will be set to `true` if new call to `VideoCapture::grab` or `VideoCapture::read` won't block.
File based `VideoCapture` instances are considered always ready.
This functions doesn't change state or dequeue any frames from `VideoCapture`, if ready `VideoCapture` passed to this function it will return immidiately.
* event - optional event which can be used to asynchronously abort waiting operation, can be null.
This function doens't change status of event, if signaled event is passed to this function it will
return immidiately.
Event must be manually reset by user.
* timeout_milliseconds - timeout in milliseconds after which this function will return. If 0 it will return immediately, which can be used for checking current `VideoCapture` status. Will wait indefinitely if negative.

#### Return value
* GRAB_ANY_TIMEOUT - timeout has expired
* GRAB_ANY_EVENT - asynchronous event was signaled, current `VideoCapture` status still updated in `items`.
* GRAB_ANY_READY - one or more `VideoCapture` instances is ready for reading.

#### Exceptions
* cv::Exception(CV_StsBadArg) - `items` is empty, `capture` field in any item is null, multiple `VideoCapture` instances have different backends.
* cv::Exception(CV_StsNotImplemented) - camera multiplexing is not supported by backend.

### `cv::GrabAnyEvent`
Events is implicitly shared
```C++
GrabAnyEvent event1;
GrabAnyEvent event2 = event1; // event2 refers to same event internally
```
`GrabAnyEvent::signal()` call is thread-safe relative to other asynchronous `GrabAnyEvent::signal()` calls.
Event state is undefined if both `GrabAnyEvent::signal()` and `GrabAnyEvent::reset()` called asynchronously.

## TODO
Check backend support for camera multiplexing without creating instances of `VideoCapturea` or `GrabAnyEvent`.

## Example

```C++
cv::VideoCapture cap0(0);
cv::VideoCapture cap1(1);

cv::GrabAnyEvent event;

...

std::vector<cv::GrabAnyItem> items(2);
items[0].capture = &cap0;
items[1].capture = &cap1;

cv::GrabAnyResult result = cv::grabAny(items, &event, 1000/*1s timeout*/);
if (result == GRAB_ANY_TIMEOUT)
{
  // Handle timeout
}
else if (result == GRAB_ANY_EVENT)
{
  // Event was signaled asynchronously
  event.reset();
}
else if (result == GRAB_ANY_READY)
{
  if (items[0].ready)
  {
    ... = cap0.read();
  }
  if (items[1].ready)
  {
    ... = cap1.read();
  }
}
```

### PR DESCRIPTION
**Pull request by Volskig**
Added static function waitAny( ..., ..., ... ) to work with multiple cameras in one thread. The method accepts a collection of VideoCapture pointers, state collection ( 1 - ready, -1 - error, 0 - not ready ), timeout. It is implemented to find the cameras states for V4L. 
Method camerasPoll( IVideoCupture --cast>> CvCupture, ..., ... ) calls function deviceHandlePoll( deviceHandle, ..., ... ). Method uses poll( ..., ..., ... ) function to find camera's state. 
User recevies a collection of states for each camera.

Proposal: [camera multiplexing API](https://github.com/opencv/opencv/issues/12077).
[Multiplex camera API demo](https://github.com/opencv/open_model_zoo/pull/291) for tests.

Device details: Intel® Core™ i7-8700 CPU @ 3.20GHz × 12. 
Cameras: Logitech C922 Pro Stream Webcam.

**FPS test:**
Parameters:
Net: Face-detection-retail-0004;
Size: Reshape (1, 3, 1000, 1000);
Launch time: 25 ms.

**Method read()**

Requests | 2 | 3 | 4 | 5 | 6 | 7
-- | -- | -- | -- | -- | -- | --
**2 cameras**| 27.1322 | 26.9473 | 26.9839 | 26.8831 | 27.0233 | -------
**3 cameras**| ------- | 27.3065 | 27.1696 | 27.2726 | 27.3191 | 27.3035
**4 cameras**| ------- | ------- | 27.2271 | 27.4342 | 26.624 | 27.0994

**Method WaitAny()**

Requests  | 2 | 3 | 4 | 5 | 6 | 7
-- | -- | -- | -- | -- | -- | --
**2 cameras**| 30.9216 | 30.7505 | 29.5278 | 29.876 | 30.8151 | -------
**3 cameras**| ------- | 29.9616 | 30.1515 | 30.8857 | 29.981 | 29.9445 
**4 cameras**| ------- | ------- | 30.69 | 30.9046 | 31.1103 | 30.5051
-------------------------------------------------------------------------------------------------
Parameters:
Net: Face-detection-adas-0001;
Size: (1, 3, 384, 672);
Launch time: 6.16 ms.

**Method read()**

Requests | 2 | 3 | 4 | 5 | 6 | 7
-- | -- | -- | -- | -- | -- | --
**2 cameras**| 81.3524 |84.9231 | 83.3405 | 85.006 | 85.1356 | -------
**3  cameras**| ------- | 86.8687 | 84.2587 | 87.7585 | 84.6507 | 84.6306
**4  cameras**| ------- | ------- | 88.7628 | 85.0765 | 89.5895 | 85.2633

**Method WaitAny()**

Requests | 2 | 3 | 4 | 5 | 6 | 7
-- | -- | -- | -- | -- | -- | --
**2 cameras**| 85.8336 | 89.1905 | 86.3272 | 85.7945 | 89.0111 | -------
**3 cameras**| ------- | 86.8968 | 84.0484 | 88.4938 | 85.3782 | 88.2053
**4 cameras**| ------- | ------- | 83.4528 | 87.8218 | 88.0895 | 87.1574

### CODE REVIEW

File: modules/videoio/src/cap_interface.hpp [hunk 1]
```diff
@@ -89,6 +96,21 @@ class LegacyCapture : public IVideoCapture
...
         return cap ? cvGrabFrame(cap) != 0 : false;
     }
+
+   bool camerasPoll(const std::vector<IVideoCapture * >& deviceHandles, std::vector<int>& state, const int64_t & timeout) CV_OVERRIDE
+   {
+       std::vector< CvCapture * > capPtrs;
+
+       for (unsigned int capnum = 0; capnum < deviceHandles.size(); ++capnum)
+       {
+           assert(dynamic_cast< LegacyCapture * >(deviceHandles[capnum]) != nullptr);
```
Comment (mshabunin) [inline modules/videoio/src/cap_interface.hpp]: Please use `CV_Assert` or `CV_DbgAssert` instead of `assert`

File: modules/videoio/test/test_camera.cpp [hunk 2]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
 
     capture.release();
 }
+TEST(COUNT_OF_CAMERA_FRAMES, v4l_poll)
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: `TEST(DISABLED_videoio_camera, v4l_poll)`

To run it call test application like this: 
```.sh
opencv_test_videoio --gtest_also_run_disabled_tests --gtest_filter=*v4poll
```

File: modules/videoio/test/test_camera.cpp [hunk 3]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+        VideoCapture cap2(2);
+
+        std::vector<VideoCapture> VCM;
+
+        VCM.push_back(cap1);
+        VCM.push_back(cap2);
+
+        std::vector<int> State;
+
+        std::vector<int> CountOfStates(2, 0);
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: It is better to rename variables to match prevailing naming style: `state`, `countOfStates`

File: modules/videoio/test/test_camera.cpp [hunk 4]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+
+        int iteration = 500;
+        for(int i = 0; i < iteration; ++i)
+        {
+            cv::VideoCapture::waitAny(VCM, State, 50);
+            for(unsigned int j = 0; j < State.size(); ++j)
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: I think it is better to use `VCM.size()` everywhere and add a check that `State.size()` is equal to it:
```.cpp
waitAny(...);
EXPECT_EQ(VCM.size(), State.size());
for (...; j < VCM.size(); ...)
```

File: modules/videoio/test/test_camera.cpp [hunk 5]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+        std::vector<VideoCapture> VCM;
+
+        VCM.push_back(cap1);
+        VCM.push_back(cap2);
+
+        std::vector<int> State;
+
+        std::vector<int> CountOfStates(2, 0);
+
+        int iteration = 500;
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: Numeric constants like this can be declared at the beginning of the test:
```.cpp
const unsigned ITERATION_COUNT = 500;
const unsigned TIMEOUT = 50;
```

File: modules/videoio/test/test_camera.cpp [hunk 6]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
 
     capture.release();
 }
+TEST(COUNT_OF_CAMERA_FRAMES, v4l_poll)
+{
+        VideoCapture cap1(0);
+        VideoCapture cap2(2);
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: It would be interesting to check how does sync works with high and low FPS camera setting:
```.cpp
// different FPS
EXPECT_TRUE(cap1.set(CAP_PROP_FPS, 5));
EXPECT_TRUE(cap2.set(CAP_PROP_FPS, 60));
// both low
EXPECT_TRUE(cap1.set(CAP_PROP_FPS, 5));
EXPECT_TRUE(cap2.set(CAP_PROP_FPS, 5));
// both high values
EXPECT_TRUE(cap1.set(CAP_PROP_FPS, 60));
EXPECT_TRUE(cap2.set(CAP_PROP_FPS, 60));
```

File: modules/videoio/test/test_camera.cpp [hunk 7]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+        int iteration = 500;
+        for(int i = 0; i < iteration; ++i)
+        {
+            cv::VideoCapture::waitAny(VCM, State, 50);
+            for(unsigned int j = 0; j < State.size(); ++j)
+            {
+                Mat new_frame1;
+                if(State[j] == 1)
+                {
+                   //if(VCM[j].retrieve(new_frame1))
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: Shouldn't it always be able to retrieve the frame if `State[j]` is 1? Then you can check it:
```.cpp
EXPECT_TRUE(VCM[j].retrieve(frame));
```

File: modules/videoio/test/test_camera.cpp [hunk 8]
```diff
@@ -104,5 +104,42 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+TEST(COUNT_OF_CAMERA_FRAMES, v4l_poll)
+{
+        VideoCapture cap1(0);
+        VideoCapture cap2(2);
+
+        std::vector<VideoCapture> VCM;
+
+        VCM.push_back(cap1);
+        VCM.push_back(cap2);
+
```
Comment (mshabunin) [inline modules/videoio/test/test_camera.cpp]: Add some checks to be sure that captures are open:
```.cpp
ASSERT_TRUE(cap1.isOpened());
ASSERT_TRUE(cap2.isOpened());
```

File: modules/videoio/perf/perf_input.cpp [hunk 9]
```diff
@@ -32,4 +33,71 @@ PERF_TEST_P(VideoCapture_Reading, ReadFile, testing::ValuesIn(bunny_files) )
...
+    TEST_CYCLE() {
+        int out = 0;
+        for(int j = 0; j < ITERATION_COUNT; ++j)
+        {
+            VideoCapture::waitAny(VCM, state, TIMEOUT);
+            for(int i = 0; i < 2; ++i)
+            {
+                if(state[i] == CAP_CAM_READY)
+                {
+                  EXPECT_TRUE(VCM[i].retrieve(forMAt[i]));
```
Comment (mshabunin) [inline modules/videoio/perf/perf_input.cpp]: Perhaps performance tests do not make much sense in this case. Standalone benchmark application (in `samples/cpp` dir) would be more useful. Lets say it could be launched in two modes: sync and async, and will perform some processing on frames (e.g. Canny) and measure performance and efficiency metrics. Take a look at the [videocapture_camera.cpp](https://github.com/opencv/opencv/blob/master/samples/cpp/videocapture_camera.cpp) sample application, it can be used as a starting point.

File: modules/videoio/test/test_camera.cpp [hunk 10]
```diff
@@ -105,4 +105,121 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
     capture.release();
 }
 
+typedef tuple<int, int, int> Size_TIME;
+typedef testing::TestWithParam< Size_TIME > DISABLED_videoio_timeout;
+
+TEST_P(DISABLED_videoio_timeout, v4l_poll_timeout)
+{
+   //Two identical cameras
+   VideoCapture cap1(0);
```
[the remark this answers is not in the document]
Comment (Volskig) [reply modules/videoio/test/test_camera.cpp]: Done

File: modules/videoio/src/cap.cpp [hunk 11]
```diff
@@ -236,6 +236,61 @@ bool VideoCapture::grab()
...
+{
+    if(!v_captures.empty())
+    {
+        int bcknd = v_captures[0].icap->getCaptureDomain();
+
+        for (unsigned int bcknd_num = 0; bcknd_num < v_captures.size(); ++bcknd_num)
+        {
+            if(!(bcknd == v_captures[bcknd_num].icap->getCaptureDomain()))
+            {
+                CV_Error(Error::StsError, "Backend equality expected");
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: Something like "All captures must have same backend"

File: modules/videoio/src/cap.cpp [hunk 12]
```diff
@@ -236,6 +236,61 @@ bool VideoCapture::grab()
...
+            }
+        }
+
+        State.resize(v_captures.size());
+
+        if(v_captures[0].waitAnyInterior(v_captures, State, timeout_nanosec))
+            return true;
+    }
+    else
+        CV_Error(Error::StsError, "Vector is empty");
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: Something like "capture list is empty"

File: modules/videoio/src/cap.cpp [hunk 13]
```diff
@@ -236,6 +236,61 @@ bool VideoCapture::grab()
     return ret;
 }
 
+bool VideoCapture::waitAny(std::vector<VideoCapture>& v_captures, std::vector<int>& State, const int & timeout_nanosec)
+{
+    if(!v_captures.empty())
+    {
+        int bcknd = v_captures[0].icap->getCaptureDomain();
+
+        for (unsigned int bcknd_num = 0; bcknd_num < v_captures.size(); ++bcknd_num)
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: use size_t 

from 1

File: modules/videoio/src/cap.cpp [hunk 14]
```diff
@@ -236,6 +236,61 @@ bool VideoCapture::grab()
...
+
+        if(v_captures[0].waitAnyInterior(v_captures, State, timeout_nanosec))
+            return true;
+    }
+    else
+        CV_Error(Error::StsError, "Vector is empty");
+    return false;
+}
+
+bool VideoCapture::waitAnyInterior(std::vector<VideoCapture>& v_captures, std::vector<int>& state,  const int & timeout)
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: IMO better to remove this function and move this logic into `waitAny`

File: modules/videoio/src/cap_interface.hpp [hunk 15]
```diff
@@ -20,6 +20,12 @@ struct CvCapture
     virtual double getProperty(int) const { return 0; }
     virtual bool setProperty(int, double) { return 0; }
     virtual bool grabFrame() { return true; }
+
+    virtual int getDeviceHandle() { return 0; }
```
Comment (Hardcode84) [inline modules/videoio/src/cap_interface.hpp]: Device handles are implementation detail, remove from common interface

File: modules/videoio/src/videoio_c.cpp [hunk 16]
```diff
@@ -70,6 +70,11 @@ CV_IMPL int cvGrabFrame(CvCapture* capture)
     return capture ? capture->grabFrame() : 0;
 }
 
+CV_IMPL bool cvCamerasPoll(const std::vector<CvCapture*>& pointers, std::vector<int>& state, const int64_t & timeout)
+{
+    return pointers[0] ? pointers[0]->camerasPoll(pointers, state, timeout) : 0;
```
Comment (Hardcode84) [inline modules/videoio/src/videoio_c.cpp]: `false` instead of 0

File: modules/videoio/test/test_camera.cpp [hunk 17]
```diff
@@ -105,4 +105,121 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+
+   std::vector<VideoCapture> VCM;
+
+   VCM.push_back(cap1);
+   VCM.push_back(cap2);
+
+   Mat frame1;
+
+   //false start
+   cap1>>frame1;
```
Comment (Hardcode84) [inline modules/videoio/test/test_camera.cpp]: Test for read success in all tests

File: modules/videoio/perf/perf_input.cpp [hunk 18]
```diff
@@ -32,4 +33,71 @@ PERF_TEST_P(VideoCapture_Reading, ReadFile, testing::ValuesIn(bunny_files) )
...
+    VideoCapture cap1(0);
+    VideoCapture cap2(2);
+
+    ASSERT_TRUE(cap1.isOpened());
+    ASSERT_TRUE(cap2.isOpened());
+
+    Mat frame1;
+    int ITERATION_COUNT = 500;
+
+    //false start
```
Comment (Hardcode84) [inline modules/videoio/perf/perf_input.cpp]: Why is this needed? Write some meaningful comment

**Comment by Hardcode84** [comment]
Need test for "waitAny"  with single VIdeoCapture.
Also nice to have tests for errors.

File: modules/videoio/src/cap.cpp [hunk 19]
```diff
@@ -236,6 +236,61 @@ bool VideoCapture::grab()
...
+
+        for (unsigned int bcknd_num = 0; bcknd_num < v_captures.size(); ++bcknd_num)
+        {
+            if(!(bcknd == v_captures[bcknd_num].icap->getCaptureDomain()))
+            {
+                CV_Error(Error::StsError, "Backend equality expected");
+            }
+        }
+
+        State.resize(v_captures.size());
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: Instead of resizing better just check size and bail with error on mismatch

File: modules/videoio/src/cap_interface.hpp [hunk 20]
```diff
@@ -89,6 +94,21 @@ class LegacyCapture : public IVideoCapture
     {
         return cap ? cvGrabFrame(cap) != 0 : false;
     }
+
+   bool camerasPoll(const std::vector<IVideoCapture * >& deviceHandles, std::vector<int>& state, const int64_t & timeout) CV_OVERRIDE
+   {
+       std::vector< CvCapture * > capPtrs;
+
+       for (size_t capnum = 0; capnum < deviceHandles.size(); ++capnum)
```
Comment (Hardcode84) [inline modules/videoio/src/cap_interface.hpp]: Use range for

File: modules/videoio/src/cap_interface.hpp [hunk 21]
```diff
@@ -89,6 +94,21 @@ class LegacyCapture : public IVideoCapture
...
+       {
+           CV_Assert(dynamic_cast< LegacyCapture * >(deviceHandles[capnum]) != nullptr);
```
Comment (Hardcode84) [inline modules/videoio/src/cap_interface.hpp]: replace with CV_DbgAssert

File: modules/videoio/src/cap_v4l.cpp [hunk 22]
```diff
@@ -344,6 +346,10 @@ struct CvCaptureCAM_V4L CV_FINAL : public CvCapture
     virtual bool grabFrame() CV_OVERRIDE;
     virtual IplImage* retrieveFrame(int) CV_OVERRIDE;
 
+    virtual bool deviceHandlePoll(const std::vector<int>&, std::vector<int>&, const int64_t &) CV_OVERRIDE;
```
Comment (Hardcode84) [inline modules/videoio/src/cap_v4l.cpp]: Pass integers by value

File: modules/videoio/src/cap_v4l.cpp [hunk 23]
```diff
@@ -881,6 +887,109 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
 
+bool CvCaptureCAM_V4L::deviceHandlePoll(const std::vector<int>& deviceHandles, std::vector<int>& state, const int64_t & timeout)
+{
+     if(!deviceHandles.empty())
+     {
+        const auto poll_flags = POLLIN | POLLRDNORM | POLLERR;
+
+        std::vector<pollfd> fds;
+
+        for (size_t dhand_num = 0; dhand_num < deviceHandles.size(); ++dhand_num)
```
Comment (Hardcode84) [inline modules/videoio/src/cap_v4l.cpp]: Range for

File: modules/videoio/test/test_camera.cpp [hunk 24]
```diff
@@ -105,4 +105,128 @@ TEST(DISABLED_videoio_camera, v4l_read_framesize)
...
+
+}
+
+typedef tuple<int, int> Size_FPS;
+typedef testing::TestWithParam< Size_FPS > DISABLED_videoio_fps;
+
+TEST_P(DISABLED_videoio_fps, v4l_poll_fps)
+{
+    //Two identical cameras
+    VideoCapture cap1(0, CAP_V4L);
```
Comment (Hardcode84) [inline modules/videoio/test/test_camera.cpp]: Unhardcode camera indices

File: modules/videoio/src/cap_v4l.cpp [hunk 25]
```diff
@@ -881,6 +887,109 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+{
+    if(!pointers.empty())
+    {
+        std::vector<int> deviceHandles;
+
+        for (size_t ptr_num = 0; ptr_num < pointers.size(); ++ptr_num)
+        {
+            if(pointers[ptr_num])
+            {
+                CV_Assert(dynamic_cast< CvCaptureCAM_V4L * >(pointers[ptr_num]) != nullptr);
```
Comment (Hardcode84) [inline modules/videoio/src/cap_v4l.cpp]: CV_DbgAssert

File: modules/videoio/src/cap_interface.hpp [hunk 26]
```diff
@@ -20,6 +20,10 @@ struct CvCapture
     virtual double getProperty(int) const { return 0; }
     virtual bool setProperty(int, double) { return 0; }
     virtual bool grabFrame() { return true; }
+    virtual bool deviceHandlePoll(const std::vector<int>&, std::vector<int>&, const int64_t &){ return false; }
+    virtual bool setFirstCapture() {return false;}
```
Comment (Hardcode84) [inline modules/videoio/src/cap_interface.hpp]: remove from public interface

File: modules/videoio/src/cap_interface.hpp [hunk 27]
```diff
@@ -20,6 +20,10 @@ struct CvCapture
     virtual double getProperty(int) const { return 0; }
     virtual bool setProperty(int, double) { return 0; }
     virtual bool grabFrame() { return true; }
+    virtual bool deviceHandlePoll(const std::vector<int>&, std::vector<int>&, const int64_t &){ return false; }
```
Comment (Hardcode84) [inline modules/videoio/src/cap_interface.hpp]: remove from public interface

File: modules/videoio/perf/perf_input.cpp [hunk 28]
```diff
@@ -21,6 +21,15 @@ const string bunny_files[] = {
     "highgui/video/big_buck_bunny.wmv"
 };
 
+static char* set_environment_variable()
```
Comment (Hardcode84) [inline modules/videoio/perf/perf_input.cpp]: something like "get_cameras_list"

File: modules/videoio/src/cap.cpp [hunk 29]
```diff
@@ -236,6 +236,52 @@ bool VideoCapture::grab()
...
+            return false;
+        for (size_t cupture_num = 0; cupture_num < state.size(); ++cupture_num)
+        {
+            if(state[cupture_num] == CAP_CAM_READY)
+            {
+                v_captures[cupture_num].icap->grabFrame();
+            }
+            if(state[cupture_num] == CAP_CAM_ERROR)
+            {
+                //printf("error cam no. %d \n", cupture_num);
```
Comment (Hardcode84) [inline modules/videoio/src/cap.cpp]: Remove debug code

File: modules/videoio/src/cap_v4l.cpp [hunk 30]
```diff
@@ -881,6 +886,109 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+            {
+                fds.push_back(pollfd{dhand_num, poll_flags, 0});
+            }
+            else
+                return false;
+        }
+        int ret = poll(fds.data(), fds.size(), timeout);
+
+        if(ret == -1)
+            return false;
```
Comment (Hardcode84) [inline modules/videoio/src/cap_v4l.cpp]: add perror call

File: modules/videoio/src/cap_v4l.cpp [hunk 31]
```diff
@@ -881,6 +886,119 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+            if(ptr_num)
+            {
+                CV_DbgAssert(dynamic_cast< CvCaptureCAM_V4L * >(ptr_num) != nullptr);
+                CvCaptureCAM_V4L *ptr = static_cast<CvCaptureCAM_V4L * >(ptr_num);
+                deviceHandles.push_back(ptr->deviceHandle);
+                ptr->setFirstCapture();
+            }
+            else
+            {
+                perror("CvCapture pointer is null");
```
Comment (Hardcode84) [inline modules/videoio/src/cap_v4l.cpp]: perror checks last errono. Is is not deeded there. Only after poll

File: modules/videoio/perf/perf_input.cpp (hunk 28 above)
Comment (Hardcode84) [reply modules/videoio/perf/perf_input.cpp]: ping

**Comment by Hardcode84** [comment]
+ Add url to original issue into PR description

**Comment by Volskig** [comment]
Tests passed.

File: modules/videoio/perf/perf_input.cpp [hunk 32]
```diff
@@ -32,4 +41,402 @@ PERF_TEST_P(VideoCapture_Reading, ReadFile, testing::ValuesIn(bunny_files) )
...
+            path.clear();
+            if(datapath_dir[step] != '\0')
+                ++step;
+        }
+        if(datapath_dir[step] == '\0')
+            break;
+        path += datapath_dir[step];
+        ++step;
+    }
+    Mat frame1, frame2,  processed1, processed2;
```
Comment (mshabunin) [inline modules/videoio/perf/perf_input.cpp]:
```.cpp
ASSERT_EQUAL(VCM.size(), 2u);
Mat frame[2], processed[2];
```

File: modules/videoio/src/cap_v4l.cpp [hunk 33]
```diff
@@ -881,6 +886,112 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+}
+
+bool CvCaptureCAM_V4L::camerasPoll(const std::vector<CvCapture*>& pointers, std::vector<int>& state, int64_t timeout)
+{
+    if(!pointers.empty())
+    {
+        std::vector<int> deviceHandles;
+        for(const auto& ptr_num : pointers)
+        {
+            if(ptr_num)
```
Comment (mshabunin) [inline modules/videoio/src/cap_v4l.cpp]: Why didn't not you use `dynamic_cast`?
```.cpp
CvCaptureCAM_V4L *ptr = dynamic_cast< CvCaptureCAM_V4L * >(ptr_num);
if (!ptr)
    return false;
deviceHandles.push_back(ptr->deviceHandle);
ptr->setFirstCapture();
```

File: modules/videoio/src/cap_v4l.cpp [hunk 34]
```diff
@@ -881,6 +886,112 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+        for (size_t struct_num = 0; struct_num < fds.size(); ++struct_num)
+        {
+            if (ret != 0)
+            {
+                if((fds[struct_num].revents & (POLLIN | POLLRDNORM)) != 0)
+                {
+                    state[struct_num] = CAP_CAM_READY;
+                }
+                else
+                    if((fds[struct_num].revents & POLLERR) != 0)
```
Comment (mshabunin) [inline modules/videoio/src/cap_v4l.cpp]: Please balance braces, its either
```.cpp
if ()
    foo();
else
    bar();
```
either
```.cpp
if ()
{
    foo();
}
else
{
    bar();
}
```

File: modules/videoio/src/cap_v4l.cpp [hunk 35]
```diff
@@ -881,6 +886,112 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
...
+        }
+        CV_DbgAssert(dynamic_cast< CvCaptureCAM_V4L * >(pointers[0]) != nullptr);
+        CvCaptureCAM_V4L *ptr = static_cast<CvCaptureCAM_V4L * >(pointers[0]);
+        if(ptr->deviceHandlePoll(deviceHandles, state, timeout))
+                return true;
+    }
+    return false;
+}
+
+bool CvCaptureCAM_V4L::setFirstCapture()
```
Comment (mshabunin) [inline modules/videoio/src/cap_v4l.cpp]: What does this method do? There is duplicated code in `grabFrame`. Duplicated code is bad practice.

File: modules/videoio/src/cap_v4l.cpp [hunk 36]
```diff
@@ -881,6 +886,112 @@ bool CvCaptureCAM_V4L::tryIoctl(unsigned long ioctlCode, void *parameter) const
     return true;
 }
 
+bool CvCaptureCAM_V4L::deviceHandlePoll(const std::vector<int>& deviceHandles, std::vector<int>& state, int64_t timeout)
+{
+     if(!deviceHandles.empty())
```
Comment (mshabunin) [inline modules/videoio/src/cap_v4l.cpp]: Please fix indents in this function.
